### PR TITLE
#318 Edit button on application home

### LINKED ISSUE #318: Show edit option on start page for other completed sections

### PR DESCRIPTION
**Pull request by nmadruga**
Fixes #318 

### Branched from
- ~~PR #384~~ MERGED
- ~~#375~~ MERGED 

### Changes
- Add edit button on `ApplicationHome` to display on application that editable, completed and before a strict section.
- Uses the `firstStrictSectionPage` to be displayed as the section with `Resume` button when user navigates to the home page.
- `SectionsProgress` component to be used by Review and Application to display the progress on each section - and options. The **badly** named received prop `changes` to be used by Review/Edit application and allow many different labels to the button that display on the same line as a section that needs attention. Thought about instead of doing passing a constructed a UI structure to this component that only renders cells... But just did this way now so we can discuss more the idea.

### CODE REVIEW

**Comment by andreievg** [comment]
Will check it out when #375 is merged and this one re-based

**Comment by nmadruga** [comment]
@andreievg I have based on #375 now, if you would like to have a look on this it should be already now

File: src/components/Sections/SectionsProgress.tsx [hunk 1]
```diff
@@ -22,19 +34,79 @@ const SectionsProgress: React.FC<SectionsProgressProps> = ({
     )
   }
 
-  const getProgress = ({ doneRequired, doneNonRequired, totalSum, valid }: ProgressType) => {
+  const getIsStrictSection = (sectionCode: string) =>
+    sectionCode === firstStrictInvalidPage?.sectionCode
+
+  const setIsAfterStrict = (sectionCode: string) => {
+    if (!isAfterStrict)
+      isAfterStrict = firstStrictInvalidPage === null ? true : getIsStrictSection(sectionCode)
```
Comment (andreievg) [inline src/components/Sections/SectionsProgress.tsx:42]: could just be `firstStrictInvalidPage === null || getIsStrictSection(sectionCode)`, but that's stylistic pref, not needed change

File: src/components/Sections/SectionsProgress.tsx [hunk 2]
```diff
@@ -22,19 +34,79 @@ const SectionsProgress: React.FC<SectionsProgressProps> = ({
     )
   }
 
-  const getProgress = ({ doneRequired, doneNonRequired, totalSum, valid }: ProgressType) => {
+  const getIsStrictSection = (sectionCode: string) =>
+    sectionCode === firstStrictInvalidPage?.sectionCode
+
+  const setIsAfterStrict = (sectionCode: string) => {
```
Comment (andreievg) [inline src/components/Sections/SectionsProgress.tsx:40]: I am not a big fan of doing setIsAfterSrtict.

Would just do:
```
isStrictSectoin = ...
isAfterStrict = ...
```
 and use that scoped variable in `SectionAction`. 

Since the methods are scoped all of the props are available.

File: src/components/Sections/SectionsProgress.tsx [hunk 3]
```diff
@@ -22,19 +34,79 @@ const SectionsProgress: React.FC<SectionsProgressProps> = ({
...
+              onClick={() =>
+                resumeApplication({
+                  sectionCode,
+                  pageNumber: firstStrictInvalidPage?.pageNumber || 1,
+                })
+              }
+            >
+              {strings.BUTTON_APPLICATION_RESUME}
+            </Button>
+          ) : progress?.completed && isAfterStrict ? (
```
Comment (andreievg) [inline src/components/Sections/SectionsProgress.tsx:85]: would this not be isBeforeStrict ?

File: src/components/Sections/SectionsProgress.tsx [hunk 4]
```diff
@@ -22,19 +34,79 @@ const SectionsProgress: React.FC<SectionsProgressProps> = ({
...
+            <Icon
+              name="pencil square"
+              color="blue"
+              style={{ minWidth: 100 }}
+              onClick={() => resumeApplication({ sectionCode, pageNumber: 1 })}
+            />
+          ) : null
+        ) : changes.state && getIsStrictSection(sectionCode) ? (
```
Comment (andreievg) [inline src/components/Sections/SectionsProgress.tsx:93]: This condition is quite hard to follow.

canEdit ? 
 isStrictSection ? 
     'Resume' : 
     progress?.completed && isAfterStrict ? 
           'Edit' : 
           changes.state && getIsStrictSection(sectionCode) ? 
                changes.label : 
                null


wouldn't this be more clear ?

<Grid.Column style={{ minWidth: 100, padding: 0 }} verticalAlign="middle" width={2}>
  <SectionAction {...{ sectionCode, isBeforeStrict, progress }} />
</Grid.Column>

Inside SectionAction ->

if(!canEdit) return  null

if(isStrictSection) return 'Resume'

if(progress?.completed && isBeforeStrict ) return 'Edit'

if( changes.state) return  changes.label

By the way, I dont' think the 'update' one in the home page can be a button (can't navigate to future pates). Would just be an indicator of how many changes are requested right ?

File: src/components/Sections/SectionsProgress.tsx [hunk 5]
```diff
@@ -47,17 +119,8 @@ const SectionsProgress: React.FC<SectionsProgressProps> = ({
...
-            </Grid.Column>
-            <Grid.Column style={{ minWidth: 100, padding: 0 }} width={2}>
-              {firstStrictInvalidPage !== null &&
-                sectionCode === firstStrictInvalidPage.sectionCode && (
-                  <Button color="blue" onClick={() => resumeApplication(firstStrictInvalidPage)}>
-                    {strings.BUTTON_APPLICATION_RESUME}
-                  </Button>
-                )}
-            </Grid.Column>
+            {canEdit && progress && <SectionProgress {...progress} />}
```
Comment (andreievg) [inline src/components/Sections/SectionsProgress.tsx:122]: Hmm, I think we want to show this even if can't edit (when LOQ is presented and before application is restarted

File: src/containers/Application/ApplicationHome.tsx [hunk 6]
```diff
@@ -45,6 +45,12 @@ const ApplicationHome: React.FC<ApplicationProps> = ({ structure, template }) =>
 
   if (!fullStructure || !fullStructure.responsesByCode) return <Loading />
 
+
+  const canUserEdit = () => {
+    const { status } = fullStructure.info.current as StageAndStatus
```
Comment (andreievg) [inline src/containers/Application/ApplicationHome.tsx:50]: It seem easier to just do fullStructure.info?.current === ApplicationStatus.Draft (or is there a problem with types ?).

Anyways this is stylistic comment no need to change

**Review by andreievg — CHANGES_REQUESTED** [review]
A couple of minor changes please

File: src/components/Sections/SectionsProgress.tsx (hunk 1 above)
Comment (nmadruga) [reply src/components/Sections/SectionsProgress.tsx:42]: I don't know if that's the same... I needed to know if the section is **after** the strict not if it is the strict

Comment (nmadruga) [reply src/components/Sections/SectionsProgress.tsx:42]: Ah! Oops, I understood now. Sorry. Will change to that, it looks better indeed

File: src/components/Sections/SectionsProgress.tsx (hunk 2 above)
Comment (nmadruga) [reply src/components/Sections/SectionsProgress.tsx:40]: No problem. That's tidier, will do that

Comment (nmadruga) [reply src/components/Sections/SectionsProgress.tsx:40]: Hmmm not sure how to do that Andrei, I was using that as a set since all sections are in order, all sections that are after the `strictInvalid` would be set to all sections after that one (without having to check the index - which isn't passed). So not sure how to do that without a set.

File: src/components/Sections/SectionsProgress.tsx (hunk 4 above)
Comment (CarlosNZ) [reply src/components/Sections/SectionsProgress.tsx:93]: I agree, I generally prefer the "guard clause" pattern more than the complex nested ternary operators.

File: src/components/Sections/SectionsProgress.tsx (hunk 3 above)
Comment (nmadruga) [reply src/components/Sections/SectionsProgress.tsx]: Thanks!
Done: 947ad03

File: src/components/Sections/SectionsProgress.tsx (hunk 2 above)
Comment (nmadruga) [reply src/components/Sections/SectionsProgress.tsx]: Done: 947ad03

File: src/components/Sections/SectionsProgress.tsx (hunk 4 above)
Comment (nmadruga) [reply src/components/Sections/SectionsProgress.tsx]: Nice. Thanks for having a thinking about that. Very helpful, will try to implement this solution now

File: src/components/Sections/SectionsProgress.tsx (hunk 5 above)
Comment (nmadruga) [reply src/components/Sections/SectionsProgress.tsx:115]: Yes, I wasn't sure. But to be improved when we start working on issue [Epic#13 - Application Changes Required](https://github.com/openmsupply/application-manager-web-app/issues/94)

File: src/components/Sections/SectionsProgress.tsx (hunk 4 above)
Comment (nmadruga) [reply src/components/Sections/SectionsProgress.tsx]: Done: de7aac7

File: src/containers/Application/ApplicationHome.tsx (hunk 6 above)
Comment (nmadruga) [reply src/containers/Application/ApplicationHome.tsx]: Done: edc427a

File: src/components/Sections/SectionsProgress.tsx (hunk 2 above)
Comment (andreievg) [reply src/components/Sections/SectionsProgress.tsx]: Oops sorry Nicole, that was a bad comment from me, the way you had it was good.

Your new changes should also work nicely

**Review by andreievg — APPROVED** [review]
LGTM, thanks for changes
